sdks/java/io/kinesis/src/main/java/org/apache/beam/sdk/io/kinesis/ShardReadersPool.java

```diff
@@ -159,4 +180,70 @@ ShardRecordsIterator createShardIterator(SimplifiedKinesisClient kinesis,
     return new ShardRecordsIterator(checkpoint, kinesis);
   }
 
+  /**
+   * Waits until all records read from given shardRecordsIterator are taken from
+   * {@link #recordsQueue} and acked.
+   * Uses {@link #numberOfRecordsInAQueueByShard} map to track the amount of remaining events.
+   */
+  private void waitUntilAllShardRecordsRead(ShardRecordsIterator shardRecordsIterator)
+      throws InterruptedException {
+    while (!allShardRecordsRead(shardRecordsIterator)) {
+      Thread.sleep(TimeUnit.SECONDS.toMillis(1));
+    }
+  }
+
+  private boolean allShardRecordsRead(final ShardRecordsIterator shardRecordsIterator) {
+    return numberOfRecordsInAQueueByShard.get(shardRecordsIterator.getShardId()).get() == 0;
+  }
+
+  private void readFromSuccessiveShards(final ShardRecordsIterator closedShardIterator)
+      throws TransientKinesisException {
+    List<ShardRecordsIterator> successiveShardRecordIterators = closedShardIterator
+        .findSuccessiveShardRecordIterators();
+
+    if (!successiveShardRecordIterators.isEmpty()) {
+
+      // Handle one closedShardIterator at a time to keep shardIteratorsMap,
+      // recordsQueue and numberOfRecordsInAQueueByShard in sync.
+      synchronized (this) {
+        ImmutableMap.Builder<String, ShardRecordsIterator> shardsMap = ImmutableMap.builder();
+        Iterable<ShardRecordsIterator> allShards = Iterables
+            .concat(shardIteratorsMap.values(), successiveShardRecordIterators);
+        for (ShardRecordsIterator iterator : allShards) {
+          if (!closedShardIterator.getShardId().equals(iterator.getShardId())) {
+            shardsMap.put(iterator.getShardId(), iterator);
+          }
+        }
+        shardIteratorsMap = shardsMap.build();
+        numberOfRecordsInAQueueByShard.remove(closedShardIterator.getShardId());
+
+        BlockingQueue<KinesisRecord> newRecordsQueue = new ArrayBlockingQueue<>(
+            queueCapacityPerShard * shardIteratorsMap.size());
+        BlockingQueue<KinesisRecord> previousRecordsQueue = recordsQueue;
+        recordsQueue = newRecordsQueue;
+        do {
+          try {
+            KinesisRecord record = previousRecordsQueue.poll(500, TimeUnit.MILLISECONDS);
+            if (record != null) {
+              newRecordsQueue.put(record);
+            }
+          } catch (InterruptedException e) {
+            LOG.warn("Thread was interrupted during resharding operation, stopping", e);
+            return;
+          }
+        } while (!previousRecordsQueue.isEmpty());
+      }
+      for (final ShardRecordsIterator recordsIterator : successiveShardRecordIterators) {
```

Comment:
This seems duplicated with code that does the same for the initial shards. Maybe extract into a method?